### PR TITLE
fix: changing node reachability status

### LINKED ISSUE #2693: node reachability status is changing
⚠️ Support requests in an issue-format will be closed immediately. For support, go to Swarm's [Discord](https://discord.gg/XGNBEKktkD).

### Context 

The libp2p reachability service appears to be emitting contradictory messages
### Expected behavior

reachability should be set and never change
### Actual behavior
it does change:
```
TRAC[2021-11-29T12:29:36Z] kademlia: connector took 170.223µs to finish: old depth 9; new depth 9 
DEBU[2021-11-29T12:29:50Z] reachability changed to Private              
INFO[2021-11-29T12:29:50Z] kademlia: updated reachability to Private    
DEBU[2021-11-29T12:29:50Z] reachability changed to Private              
INFO[2021-11-29T12:29:50Z] kademlia: updated reachability to Private    
DEBU[2021-11-29T12:29:51Z] batch service: updated block height to 19319040 
TRAC[2021-11-29T12:29:51Z] kademlia: connector took 184.536µs to finish: old depth 9; new depth 9 
TRAC[2021-11-29T12:30:01Z] updated exchange rate to 100000 and deduction to 100 
TRAC[2021-11-29T12:30:06Z] kademlia: connector took 216.791µs to finish: old depth 9; new depth 9 
INFO[2021-11-29T12:30:09Z] debug api access                              duration=0.028536936 ip=127.0.0.1 method=GET proto=HTTP/1.1 size=383 status=200 uri=/addresses user-agent="Mozilla/5.0 (X11; Linux x86_64; rv:94.0) Gecko/20100101 Firefox/94.0"
DEBU[2021-11-29T12:30:16Z] batch service: updated block height to 19319045 
TRAC[2021-11-29T12:30:21Z] kademlia: connector took 153.591µs to finish: old depth 9; new depth 9 
TRAC[2021-11-29T12:30:36Z] kademlia: connector took 233.318µs to finish: old depth 9; new depth 9 
DEBU[2021-11-29T12:30:41Z] batch service: updated block height to 19319050 
TRAC[2021-11-29T12:30:51Z] kademlia: connector took 148.273µs to finish: old depth 9; new depth 9 
INFO[2021-11-29T12:30:57Z] debug api access                              duration=0.386428757 ip=127.0.0.1 method=GET proto=HTTP/1.1 size=62 status=200 uri=/chequebook/balance user-agent="Mozilla/5.0 (X11; Linux x86_64; rv:94.0) Gecko/20100101 Firefox/94.0"
DEBU[2021-11-29T12:31:05Z] reachability changed to Public               
INFO[2021-11-29T12:31:05Z] kademlia: updated reachability to Public     
TRAC[2021-11-29T12:31:06Z] kademlia: connector took 155.23µs to finish: old depth 9; new depth 9 
DEBU[2021-11-29T12:31:06Z] batch service: updated block height to 19319055 
DEBU[2021-11-29T12:31:10Z] reachability changed to Public               
INFO[2021-11-29T12:31:10Z] kademlia: updated reachability to Public     
```

### Steps to reproduce
run a light node on mainnet

### Possible solution
set the reachability with `sync.Once`

### PR DESCRIPTION
**Pull request by mrekucci**
### Checklist

- [ ] I have read the [coding guide](https://github.com/ethersphere/bee/blob/master/CODING.md)
- [ ] My change requires a documentation update and I have done it
- [ ] I have added tests to cover my changes.

### Screenshots (if appropriate):

Closes #2693

---
This change is [Reviewable](https://reviewable.io/reviews/ethersphere/bee/2728)

### CODE REVIEW

**Review by acud — APPROVED** [review]

File: pkg/topology/kademlia/kademlia.go [hunk 1]
```diff
@@ -1316,10 +1317,13 @@ func (k *Kad) Reachable(addr swarm.Address, status p2p.ReachabilityStatus) {
 }
 
 // UpdateReachability updates node reachability status.
+// The status will be updated only once.
 func (k *Kad) UpdateReachability(status p2p.ReachabilityStatus) {
-	k.logger.Infof("kademlia: updated reachability to %s", status.String())
-	k.reachability = status
-	k.metrics.ReachabilityStatus.WithLabelValues(status.String()).Set(0)
+	k.setReachabilityOnce.Do(func() {
```
Comment (aloknerurkar) [inline pkg/topology/kademlia/kademlia.go:1322]: Not sure if this is what we want to do. I think it's possible that the reachability can change over the life of the node. Having a `sync.Once` seems too restrictive.

Instead, I would simply put an if loop around the metrics update call to update only if the value was different that the original.

Comment (mrekucci) [reply pkg/topology/kademlia/kademlia.go:1322]: But then you'll have inconsistencies between what is set in metrics and what will be reported in the topology endpoint.

**Comment by aloknerurkar** [comment]
___
*[pkg/topology/kademlia/kademlia.go, line 1322 at r1](https://reviewable.io/reviews/ethersphere/bee/2728#-MqUBwjcDyhdRzlSnkTk-r1-1322:-Mqn2qu77bjH9oYZokD6:b-ie4pkm) ([raw file](https://github.com/ethersphere/bee/blob/e722ab44abb73ed59d9c86125a4c4e9e231688a0/pkg/topology/kademlia/kademlia.go#L1322)):*
<details><summary><i>Previously, mrekucci wrote…</i></summary><blockquote>

But then you'll have inconsistencies between what is set in metrics and what will be reported in the topology endpoint.
</blockquote></details>

So I need to understand why we need the `sync.Once` here first.

**Comment by mrekucci** [comment]
> _[pkg/topology/kademlia/kademlia.go, line 1322 at r1](https://reviewable.io/reviews/ethersphere/bee/2728#-MqUBwjcDyhdRzlSnkTk-r1-1322:-Mqn2qu77bjH9oYZokD6:b-ie4pkm) ([raw file](https://github.com/ethersphere/bee/blob/e722ab44abb73ed59d9c86125a4c4e9e231688a0/pkg/topology/kademlia/kademlia.go#L1322)):_
> Previously, mrekucci wrote…
> 
> So I need to understand why we need the `sync.Once` here first.

It's basically because of this issue (in libp2p): #2693

**Review by notanatol — APPROVED** [review]

**Review by acud — COMMENTED** [review]
> Reviewed 1 of 1 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/ethersphere/bee/2728)* status: all files reviewed, 1 unresolved discussion (waiting on @aloknerurkar and @istae)

___
*[pkg/topology/kademlia/kademlia.go, line 1322 at r1](https://reviewable.io/reviews/ethersphere/bee/2728#-MqUBwjcDyhdRzlSnkTk-r1-1322:-Mqnry40AEAe09mBEjNR:bpqhtl1) ([raw file](https://github.com/ethersphere/bee/blob/e722ab44abb73ed59d9c86125a4c4e9e231688a0/pkg/topology/kademlia/kademlia.go#L1322)):*
<details><summary><i>Previously, aloknerurkar wrote…</i></summary><blockquote>

So I need to understand why we need the `sync.Once` here first.
</blockquote></details>

@aloknerurkar if the reachability is constantly changing, how do you know which value to pick?

**Comment by aloknerurkar** [comment]
> > Reviewed 1 of 1 files at r1, all commit messages.
> > _[Reviewable](https://reviewable.io/reviews/ethersphere/bee/2728)_ status: all files reviewed, 1 unresolved discussion (waiting on @aloknerurkar and @istae)
> 
> _[pkg/topology/kademlia/kademlia.go, line 1322 at r1](https://reviewable.io/reviews/ethersphere/bee/2728#-MqUBwjcDyhdRzlSnkTk-r1-1322:-Mqnry40AEAe09mBEjNR:bpqhtl1) ([raw file](https://github.com/ethersphere/bee/blob/e722ab44abb73ed59d9c86125a4c4e9e231688a0/pkg/topology/kademlia/kademlia.go#L1322)):_
> 
> Previously, aloknerurkar wrote…
> @aloknerurkar if the reachability is constantly changing, how do you know which value to pick?

So the reachability update is being provided by the libp2p event bus. If it changes, shouldnt we report the updated value?

For eg. due to some network issue, if the reachability becomes private from public, the topology endpoint should show the updated value right?

**Review by acud — APPROVED** [review]
I opened an issue on their repo about this, but still no comments: https://github.com/libp2p/go-libp2p-autonat/issues/113. If you look at the autonat code you will see that it is not really binary in the sense that it applies some heuristic with regards to determining the reachability status. If you look at their repo, you'll also see that this is entirely not trivial, so it is also not clear to which degree we can rely on this (but that is an entirely different story....) status. It is not very much clear to me at a glance what would make the status change, but even if it does, the value would need to have some easing applied over it if we want to improve reliability. Otherwise you end up with a node that one moment considers itself for routing and doesn't in the other. This can be really noisy and flaky.

> *[Reviewable](https://reviewable.io/reviews/ethersphere/bee/2728#-:-Mqp-BjiCAeP8G9TlJlV:b93aedj)* status: all files reviewed, 1 unresolved discussion (waiting on @aloknerurkar and @istae)

**Review by aloknerurkar — APPROVED** [review]

**Pull request merged**